{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Rosalind.ipynb",
      "provenance": [],
      "mount_file_id": "1-ngsQAqMLceC3BOC1YLJqk-cPoybLPv-",
      "authorship_tag": "ABX9TyMLHw35NYoMiaW31bVfTkkS",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/janerik/tools/blob/main/Rosalind.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "q3yUk4oRWULn"
      },
      "outputs": [],
      "source": [
        "## Inferring mRNA from Protein\n",
        "\n",
        "translation = \"\"\"UUU F      CUU L      AUU I      GUU V\n",
        "UUC F      CUC L      AUC I      GUC V\n",
        "UUA L      CUA L      AUA I      GUA V\n",
        "UUG L      CUG L      AUG M      GUG V\n",
        "UCU S      CCU P      ACU T      GCU A\n",
        "UCC S      CCC P      ACC T      GCC A\n",
        "UCA S      CCA P      ACA T      GCA A\n",
        "UCG S      CCG P      ACG T      GCG A\n",
        "UAU Y      CAU H      AAU N      GAU D\n",
        "UAC Y      CAC H      AAC N      GAC D\n",
        "UAA Stop   CAA Q      AAA K      GAA E\n",
        "UAG Stop   CAG Q      AAG K      GAG E\n",
        "UGU C      CGU R      AGU S      GGU G\n",
        "UGC C      CGC R      AGC S      GGC G\n",
        "UGA Stop   CGA R      AGA R      GGA G\n",
        "UGG W      CGG R      AGG R      GGG G\"\"\"\n",
        "\n",
        "coded = (\"MSNEIWTTNIESEVPWAMSYAKKMVKQFHSCWGWGASVGQCKLRIWPQIPIRDFKLKVYHYIQQTYNVPID\"\n",
        "\"WHMYQGPDWEWLKDDHAQHLNDPSLMKWAPDVMHCIYIQLIKTDNVYKCCWNFPDNCMPWYAELQCIHTRCYSRLMFRET\"\n",
        "\"ETYHWCQCMQKTAVARQYSFIKQAGRKGRPPMLQGLCSRNGRERCMSGQLGRGHYFVTNVYYQRQFMHLQSIMHEPECPK\"\n",
        "\"WLLVTYVMDLVNCHKWMGMGWKHFWIHVHRYMHGSQWLPWRRTDYAGGAQQTLNAGVTRCHHRGFWCLGTVCVLDINFRE\"\n",
        "\"WMLGNCFDAQSNHHQETYRRVKWHYCSLEGMDQDPFQEISHRPWTPSRVFPHMDSGWLYWFYGHEDYVTARRWKQLHYWE\"\n",
        "\"SQTHKYEIECMQILLQNYGKHRSQSFMRILTWHKIDQQCYHVKQRCCNGCTHERICQQVHCPWVTVHYPTRWKNCTNSHV\"\n",
        "\"TSDNLNLMMACSFVPYSQKVINPIMELFDAMWIGTFRDCQCAGCGGMVLRGQPNKKLANARHKVGPNNNTFFTHMGVWRG\"\n",
        "\"NTKRHWLMGMQKCVFKQVSHNGYYQNFVHNWGPLTQKWARLTADHTSAAIMNYVVGKMALVRFWNWGEPVDPRAIQFDCP\"\n",
        "\"MAHYQFFWAVYTRISHRDNVGYTPKLFGQQINHNSSQCVCEEWVWYTNFEPTSQIKVFQHDLDLQYQLFNSNKYPPSEKC\"\n",
        "\"SDHYFLSRMNQCPVWSIGFGMEKIPINRCDRRMWFWWCFRGGHYINMELCIRGFCPPRSDCRFVDFHLLTAEWTWKIGSL\"\n",
        "\"YQGNAGTHHDGPYLLTKDWWFMVRLMEWMTDSFCMSPRCKANPSTNTGLGGQNEVTMSLVIMFLKYLKAPAPYSPEVQNM\"\n",
        "\"SKWWDANDHFINANSIWYRKGPMVQTERRVMNHEIAQFENIFKEDSAHEDGHTCIDVCSGPSNLNTHVVWDAEMNCAFCK\"\n",
        "\"CCGEDKWLIYIKLPCAYHGICDCICREWQSCFQVCSDTTWWTCWMVVKF\")\n",
        "number = translation.count(\"Stop\")\n",
        "\n",
        "traL =  translation.split()\n",
        "list_aac = traL[1::2]\n",
        "list_aac = list(filter((\"Stop\").__ne__, list_aac))\n",
        "\n",
        "for aacid in coded:\n",
        "  number *= list_aac.count(aacid)\n",
        "\n",
        "print(number % 1000000)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## Finding a Protein Motif\n",
        "\n",
        "import urllib.request\n",
        "\n",
        "base_url = \"http://www.uniprot.org/uniprot/\"\n",
        "\n",
        "ids=[\"A9QYR8\",\n",
        "\"P10761_ZP3_MOUSE\",\n",
        "\"O08537_ESR2_MOUSE\",\n",
        "\"P22891_PRTZ_HUMAN\",\n",
        "\"P00304_ARA3_AMBEL\",\n",
        "\"A1UR66\",\n",
        "\"P31096_OSTP_BOVIN\",\n",
        "\"P37803\",\n",
        "\"Q4JAS3\",\n",
        "\"P01046_KNL1_BOVIN\",\n",
        "\"P27918_PROP_HUMAN\",\n",
        "\"P23185\",\n",
        "\"P07306_LECH_HUMAN\"\n",
        "]\n",
        "\n",
        "for id in ids:\n",
        "  locations = \"\"\n",
        "  target_url = base_url + id + \".fasta\"\n",
        "  seq = \"\"\n",
        "  for i, line in enumerate(urllib.request.urlopen(target_url)):\n",
        "    if i > 0:\n",
        "      seq += line[0:-1].decode(\"utf-8\")\n",
        "      \n",
        "  for i in range(len(seq)-4):\n",
        "    if seq[i] == \"N\" and seq[i+1]!=\"P\" and (seq[i+2]==\"S\" or seq[i+2]==\"T\") and seq[i+3]!=\"P\":\n",
        "      locations += str(i+1) + \" \"\n",
        "  if locations!=\"\":\n",
        "    print(id)\n",
        "    print(locations.strip())"
      ],
      "metadata": {
        "id": "cTVJX7yFdKPO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Open Reading Frames\n",
        "\n",
        "org = (\"GCGTTTCCCCGCAGGTAACTATACGCGTAGGTGTAAATAATTTTCGTTCACGTGTGTTAG\"\n",
        "\"TCTACGTGGCTCGTCGTGTACGGAACTAAAGTGAGTAGGATAAGAAGTAGCAAAGACAGT\"\n",
        "\"CAACACCGTCGTGGGTCCACTTCAAGTGTTTCGGTATGTCGAGCTTATACCACTATAAAC\"\n",
        "\"TAGTCAGGATCCCATTCATAGTCTCCCAGAGTGAGTAATAGGGTCGGCTTTAATTCAGTC\"\n",
        "\"GTCGGTACGATAGTGCATTAGGATGAATGTCCACGGGAGTATAGTTCCCGTGCGTGAGCG\"\n",
        "\"TGGACCACTAGTAGAGCAGCTCCGTGGAGTCTCCCTCTGGAAACCTACGAGGCCGACGGA\"\n",
        "\"TGCCCCTGCGTCCTGCCAGCAGCCAGAGTGTTATCTTTAATATGAAGCAGATCCCTTCGT\"\n",
        "\"GGACCGGCCCATAGCTATGGGCCGGTCCACGAAGGGATCTGCTTCATAGAGGAATATGAC\"\n",
        "\"ATGGTTAGGCTATCAGCGGTATTGAGATCTCACGGACACATAGACTCCCACGCAACTGGG\"\n",
        "\"CATAGCTGGTGTTTTAATTTGTGGAACCGGCGTTGGAGATATGGCTACCCTCGTTCTGAG\"\n",
        "\"GAGACAATGCCTTTCGCTTAAGAGACGCCACCAAAATGCGAGTAAGGCGCCTACTCGTGG\"\n",
        "\"TTGAACTAAAGTTGTCAATATAAGGATGGTCACAACTCAGGCCTGCAAGGGTTATAAAAC\"\n",
        "\"CTAGAGGGAAAATGGCAGGAGGTCTGCAACGGCCCGGCCGTATTCCTTCCGTCTTATGTG\"\n",
        "\"GCAATACGCGTCACCTCTCCTCCTCGGTGCCTGCTAGTTGGTTAACACGCCTTCGCAGCT\"\n",
        "\"CCTGCGTTAGGCTGTCTCAATACTAAAA\")\n",
        "reverse_comp = org[::-1]\n",
        "reverse_comp = reverse_comp.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()\n",
        "\n",
        "strands = [org, reverse_comp]\n",
        "\n",
        "translation = \"\"\"TTT F      CTT L      ATT I      GTT V\n",
        "TTC F      CTC L      ATC I      GTC V\n",
        "TTA L      CTA L      ATA I      GTA V\n",
        "TTG L      CTG L      ATG M      GTG V\n",
        "TCT S      CCT P      ACT T      GCT A\n",
        "TCC S      CCC P      ACC T      GCC A\n",
        "TCA S      CCA P      ACA T      GCA A\n",
        "TCG S      CCG P      ACG T      GCG A\n",
        "TAT Y      CAT H      AAT N      GAT D\n",
        "TAC Y      CAC H      AAC N      GAC D\n",
        "CAA Q      AAA K      GAA E\n",
        "CAG Q      AAG K      GAG E\n",
        "TGT C      CGT R      AGT S      GGT G\n",
        "TGC C      CGC R      AGC S      GGC G\n",
        "CGA R      AGA R      GGA G\n",
        "TGG W      CGG R      AGG R      GGG G \"\"\"\n",
        "\n",
        "traL =  translation.split()\n",
        "traDict = dict(zip(traL[0::2], traL[1::2]))\n",
        "protein_list = []\n",
        "\n",
        "for coded in strands:\n",
        "  for j in range(0, len(coded)-3, 1):\n",
        "    decoded = \"\"\n",
        "    start = 0\n",
        "    for i in range(j, len(coded)-3, 3):\n",
        "      if coded[i:i+3] in (\"TAA\",\"TAG\",\"TGA\"):\n",
        "        start = 0\n",
        "        break\n",
        "      if start==1:\n",
        "        decoded += traDict[coded[i:i+3]]\n",
        "        continue\n",
        "      if coded[i:i+3] == \"ATG\":\n",
        "        decoded += \"M\"\n",
        "        start = 1\n",
        "\n",
        "    if decoded != \"\" and decoded not in protein_list and start == 0:\n",
        "      protein_list.append(decoded)\n",
        "\n",
        "for protein in protein_list:\n",
        "  print(protein)"
      ],
      "metadata": {
        "id": "aj2MOX7ecUn4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Locating Restriction Sites\n",
        "\n",
        "org = (\"CGCGTGACTCGATTGAGTTTAGCCGGACACAGGCTGACTAGTCTGACCTCTACTGAGCAT\"\n",
        "\"CGTCTTATATCGTACGTTAACACGTATAGAGACTTTGTTAGTGAGCGGCGGTAGCAAGAG\"\n",
        "\"CGGAGGTATGTCTTCGGATGACGGGTTCATAATATTACATGCACGGGTAGTTCAGCGGTG\"\n",
        "\"TACGGGCCCCATCTACGGAGAACGTGGTGCGTGGGGGTAAATCCGGCGACTCGGAGACAA\"\n",
        "\"TCATGCAAAGGCCCCTGCAACGCGCCCTCGTTTGCTGTCGGATACAAAGAATTTCACTCC\"\n",
        "\"AGTTACGCCATCCCAGAACAAACTCACGCCACCTTATGGCGGATGAAGTAGACATTAGGC\"\n",
        "\"TTCCTCCCACGGAGACTAGATTGTGCTGGCGTCGTTCGTAGATGGTGACACAGGGCTAGT\"\n",
        "\"CTTCCGTCCGCATGTCTACACTACAACATGGGCGGGACGAGGCGCACGTGCGCGTGGTTT\"\n",
        "\"TTGTGGTAGCATGTCCACCAAGCGGCTGCGCGATATTCATAATCTCCTTCCGGACGCACC\"\n",
        "\"AACTAGCGATAATGTGCGCTTGATGCCGAGAAATGAGTCGTACCGTTGCTTGATCAGAAT\"\n",
        "\"ATCATTGTGACGCGCTATTAAAGCACTACCTCCTGTTTAAAGCCCGTCGTCCACCTATTG\"\n",
        "\"AATCATAGCTGTGACGTTGCAATTGGGGTACCGTACTGGGGCATCACGGTGGCACTACCG\"\n",
        "\"TCCCGCTCACCAGCGAGGTGAGGCCGGCTTCGCCTATCGCACAATCAATAATTTGTTCTT\"\n",
        "\"CCAGAAAATTGAGCAGGCTAAAACCCCTGGTCCAGGTGGCAACCGACGGGCGGATGTTGC\"\n",
        "\"CCTTAGGCGGTACTTGGAACGTGACCACCGGTCTGAATCGTGCTCAGAGCCGGATGGAGC\"\n",
        "\"CGCGAGTAGTGCGATCGGGTGCCTAACGGAGCGC\")\n",
        "list_palindromes = []\n",
        "locations = []\n",
        "\n",
        "for i in range(12,3,-2):\n",
        "  for j in range(0,len(org)-i+1):\n",
        "    reverse_comp = org[j:j+i]\n",
        "    reverse_comp = reverse_comp.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()[::-1]\n",
        "    if org[j:j+i] == reverse_comp:\n",
        "      locations.append(j+1)\n",
        "      list_palindromes.append([j+1,i])\n",
        "\n",
        "for item in list_palindromes:\n",
        "  print('\\t'.join(str(e) for e in item))"
      ],
      "metadata": {
        "id": "24NQcT0HL-4f"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "##  Transitions and Transversions\n",
        "\n",
        "seq1 = (\"ATAGGCAACTGGAAAAAGCAACAGGCGGCATCACTATTATCTTCAGAGCCTTTAAACGTA\"\n",
        "\"AGGTCCGCAGGCCGCGAGCGACCCGTAGGTCCCAATCGAATTTAGATCTAGTCAGATCCC\"\n",
        "\"TGCGTCCAGAAAAATCATTACACTAGAAAGGCATGGACCTATATAGACGGATGATCGAAA\"\n",
        "\"CATCGGCTGGTCAGGTACGCTAACCAGTCTTTGGTGTCGTCTGAAAGCATGCCCTGTCCC\"\n",
        "\"TATAACTTCCCCGCTCCAAATTTTTACGCCCGGGATGGGCCATGTTACTCGCCTCGGTCC\"\n",
        "\"TTGTATCTAGGTCAGTTGTACGGCGCCGTAGCAGCGAAGACCAGGAGTTGTTTTCCATAT\"\n",
        "\"AGGGCCTCACTAGGCAAACCAATTACTATGTTGCGTCTGCTACCCTACTGCACCAAAAGC\"\n",
        "\"AAATGCGCACGGCTCTTTAGCGATAACCAGCTGGCGTGTCGGAGCTCCTCCGAATATTCA\"\n",
        "\"CCGCTGTACAGCCTGATAAGTACTCAAACAGTGGGTATATTTGGGGTTGACATCATGTTC\"\n",
        "\"TACACTAACGCGCGCCATGCTGCGTCAGCAGATAACTTGCACGATGTAACTATGTTGACT\"\n",
        "\"ATACTAGCTTACGGAACATTAGTGTATATGGAGGGAACCTATTCGAGCCGGTTTTCGTTA\"\n",
        "\"TGTTAACCGGTCGTGGGGGACAACTTAACACTCATTATAGCGAGGTAACGGTAATGTTCC\"\n",
        "\"AAGGCTCCAGGATCGGAGTAGGCCTGGCTTGTGCTCCCGTTCGCTGCTGGTCACCACTCA\"\n",
        "\"AATATTCACGGTTGATCATA\")\n",
        "seq2 = (\"ATAAGTTGCTGGGAAAAGCATGAAGCGACATCAATATCACTCTCGGAATCCTCACTCGCA\"\n",
        "\"AGATACGCAGGCCCCGGGCGACCGGTAAGTCCTGATCGAATTCAAATTCGGTCGAATCTG\"\n",
        "\"TGCCTCTGGAAAACCCATTTTGCTAAAAAGGCACCGATGTAGATAGATTGCTAATCGTAA\"\n",
        "\"CATTAGTTGGTCAGGCAGGATCGCCAATCGTTGGTGTTCTTTGAAAGAATGGCCTATCAC\"\n",
        "\"TAAAACTCGTCTGCCCCAAATATGTATAACCGTGACGAATGAGGTTACCCGCCTCGGCCC\"\n",
        "\"TCTTACTCAGGGCAGTCGTGCAGCGCAACAGGAACGAAGCCAGAGAGTCGTTCTCTATGT\"\n",
        "\"GGGGGTCTATTAGATGAACCAAGTACCACGCCACATCTCCCGTTTTACCGCGCCAAACGC\"\n",
        "\"AAATGCGCATGGCTCTTGAACAATAACTGAATGGCCTGCCGTAGCCCTCCAGAAGATTCG\"\n",
        "\"CCGCTACCCAATGTAGTGAGTACTCAATCAGTAGATAACCTTTGGTATGGGATCATGCTC\"\n",
        "\"TACATTTGCGAATGTCATATGGGCTCAGCACAAAAGCTTCACGACGCAACTATATGGTCT\"\n",
        "\"CTTCAAATTCACGGTGCATTAGTGTATGTGGAAGAAACTTATTCCGGCTGCTTTTCGTTA\"\n",
        "\"TGCGAACCTATCATGTGGGGCAACTTGCCCCTCATTATGACTAAATAACGGTAATGTCCC\"\n",
        "\"AAGACTCCAGAATAGGAATAGGTCCGGTAGGGGGTCCCGTCTGCCGTTAGTCACCACTCT\"\n",
        "\"AGCACACACGGTTAGTCATA\")\n",
        "\n",
        "transitions = 0\n",
        "transversion = 0\n",
        "\n",
        "for i, base in enumerate(seq1):\n",
        "  if base == seq2[i]:\n",
        "    continue\n",
        "  elif (base == \"A\" and seq2[i] == \"G\") or (base == \"G\" and seq2[i] == \"A\") or (base == \"C\" and seq2[i] == \"T\") or (base == \"T\" and seq2[i] == \"C\"):\n",
        "    transitions += 1\n",
        "  else:\n",
        "    transversion += 1\n",
        "\n",
        "print(round(transitions/transversion,11))"
      ],
      "metadata": {
        "id": "bDcVTVAYXaNs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "##  Finding a Spliced Motif\n",
        "\n",
        "import re\n",
        "\n",
        "seq1 = (\"CAAGTATTCACTTAGTCTGAGTACGAAACCTAATAGAGGTCTCCAAAGCTAAACTCATGG\"\n",
        "\"CATCGAGTGTTCGTACAGGATGATGCGGGAGAGCGACCTCTGTTGCGAAACGAATTTGTT\"\n",
        "\"TTGTTTGTCACAGGGTAGATTCGCTCTTGAGGTATGCGGGCTGGAATAGCGGGTACGTTC\"\n",
        "\"GGTACATCGGGTCAGGGACCCTACCATACAGTCTCTAGCGTCATTACTCTGAAAACGAAA\"\n",
        "\"ACCCATTGTGCGGGAGTGTGTTCTCGGCATACGAACTGGGAAACTAAGATGATAATTTAA\"\n",
        "\"GCACATTCCATAACATAAGTATTCCTAGACATGAGCTTCTAATAATCGCTGGTTCCTGGT\"\n",
        "\"CTAACTTTGCAAAGGCCTACGGTTCGCTTATAAATGCATGCCGGAATGAGTGCGGATTGG\"\n",
        "\"CGCGTGCGGGTTGGTCGGAAATAACTGGGACCTCTAGATTCGAGGTGATGTTGGCGCCCT\"\n",
        "\"ACTGCCCCGTCCTTCAGCGTTCGCTCTACTATAGGACACCCAGTGTCAAGAATCGAGTAG\"\n",
        "\"CCTGTTCTAATCAGCACGTAGATCGGGGAACTTTTCGACGGATCGCCGTCCAACCTTCTT\"\n",
        "\"TCACCGGCTAAGTTCTAGGCGAGGAATCAGATATCAACATACGATCAGAGAATTATCTAC\"\n",
        "\"ATGGTTTTGCGTTAGCGAATTTGATTCACGTGGTGTTCCTCCTTCGGCGTGGATACAATC\"\n",
        "\"TCAGCTTGAGGCACTCAAGGCAACCGGTCCATTATCGTAATCGGTCGGTTGGTTGCTACT\"\n",
        "\"ACAGGTCCTGAATCGCAATCCCGGTCGAGGAACTACGAACGCTACGCGACGAGACGCGCG\"\n",
        "\"TTACCTCGAGGCTTTCGACCTTACATTTCGAAGCCTCAAAATTCAAATTGCTGATAGTCTCGGGC\")\n",
        "seq2 = \"AAGAGGGGTACTCACTAACG\"\n",
        "\n",
        "indices = [0]\n",
        "\n",
        "for letter in seq2:\n",
        "  index = seq1.find(letter,indices[-1]+1)\n",
        "  indices.append(index)\n",
        "\n",
        "indices.pop(0)\n",
        "print(' '.join(str(e+1) for e in indices))"
      ],
      "metadata": {
        "id": "95beouEzaKk2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Bio"
      ],
      "metadata": {
        "id": "14NZSvZIfMvE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Genome Assembly as Shortest Superstring\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "file_in =\"/content/drive/MyDrive/Rosalind/rosalind_long.txt\"\n",
        "sequence = \"\"\n",
        "\n",
        "records = list(SeqIO.parse(file_in, \"fasta\"))\n",
        "sequences = []\n",
        "for record in records:\n",
        "  sequences.append(record.seq)\n",
        "sequence = records[0].seq\n",
        "done = 1\n",
        "sequence_nr = 0\n",
        "while done < len(records):\n",
        "  alignment = 0\n",
        "  back = False\n",
        "  front = False\n",
        "  sequences.pop(sequence_nr)\n",
        "  sequence_nr = 0\n",
        "  for i in range(len(sequences)):\n",
        "    curr_seq = sequences[i]\n",
        "    if sequence.find(curr_seq) == -1:\n",
        "      for j in range(len(curr_seq),0,-1):\n",
        "        if sequence[-j:] == curr_seq[0:j] and j > alignment:\n",
        "          alignment = j\n",
        "          back = True\n",
        "          front = not back\n",
        "          sequence_nr = i\n",
        "          break\n",
        "        if sequence[0:j] == curr_seq[-j:] and j > alignment:\n",
        "          alignment = j\n",
        "          back = False\n",
        "          front = True\n",
        "          sequence_nr = i\n",
        "          break\n",
        "    else:\n",
        "      done += 1\n",
        "      sequence_nr = i\n",
        "      continue\n",
        "\n",
        "  if back == True:\n",
        "    sequence += (sequences[sequence_nr])[alignment:]\n",
        "  elif front == True:\n",
        "    sequence = (sequences[sequence_nr])[0:-alignment] + sequence\n",
        "  \n",
        "  done +=1\n",
        "\n",
        "print(sequence)"
      ],
      "metadata": {
        "id": "xi4eRZoQe9gW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Error Correction in Reads\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "file_in =\"/content/drive/MyDrive/Rosalind/rosalind_corr.txt\"\n",
        "records = list(SeqIO.parse(file_in, \"fasta\"))\n",
        "orig = []\n",
        "for record in records:\n",
        "  orig.append(record.seq)\n",
        "\n",
        "all = {}\n",
        "\n",
        "def make_reverse_comp(sequence_input):\n",
        "  reverse_comp = sequence_input.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()[::-1]\n",
        "  return reverse_comp\n",
        "\n",
        "def hamming_dist(seq1,seq2):\n",
        "  distance = 0\n",
        "  for i in range(len(seq1)):\n",
        "    if seq1[i]!=seq2[i]:\n",
        "      distance += 1\n",
        "  return distance\n",
        "\n",
        "for record in records:\n",
        "  if record.seq in all:\n",
        "    all[record.seq] += 1\n",
        "    all[make_reverse_comp(record.seq)] += 1\n",
        "  else:\n",
        "    all[record.seq] = 1\n",
        "    all[make_reverse_comp(record.seq)] = 1\n",
        "\n",
        "for item in all:\n",
        "  if all[item]==1:\n",
        "    for vergl in all:\n",
        "      if hamming_dist(item,vergl)==1 and all[vergl]>1 and item in orig:\n",
        "        print(f'{item}->{vergl}')\n",
        "        break"
      ],
      "metadata": {
        "id": "-L16dKWYpCas"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Enumerating Oriented Gene Orderings\n",
        "\n",
        "def permutation(lst):\n",
        "    if len(lst) == 0:\n",
        "        return []\n",
        "    if len(lst) == 1:\n",
        "        return [lst]\n",
        " \n",
        "    l = []\n",
        "    for i in range(len(lst)):\n",
        "       m = lst[i]\n",
        "       remLst = lst[:i] + lst[i+1:]\n",
        "       for p in permutation(remLst):\n",
        "           l.append([m] + p)\n",
        "    return l\n",
        "\n",
        "n = 4\n",
        "vector = [i+1 for i in range(n)]\n",
        "\n",
        "perms = permutation(vector)\n",
        "\n",
        "for i in range(len(vector),0,-1):\n",
        "  negative = [-1 for i in range(i)]\n",
        "  negative.extend([1 for i in range(len(vector)-i)])\n",
        "  for perm in permutation(negative):\n",
        "    new_perms = permutation([a*b for a,b in zip(perm,vector)])\n",
        "    if new_perms[0] not in perms:\n",
        "      perms.extend(new_perms)\n",
        "\n",
        "print(len(perms))\n",
        "for item in perms:\n",
        "  print(' '.join(str(e) for e in item))"
      ],
      "metadata": {
        "id": "3QX4-nDvNoAR"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}